### PR TITLE
Feature/rule macro

### PR DESCRIPTION
**Pull request by michel-steuwer**
Added macros `@rule`, `@strategy` and `@combinator` to avoid boiler plate code.

### CODE REVIEW

**Review by johanneslenfers — DISMISSED** [review]
like it

File: macros/src/main/scala/elevate/macros/CombinatorMacro.scala [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+          } else {
+            q"def $name[..$tparams](...$regularParamLists)(...$funImplParamLists): $returnType = $body"
+          }
+        }
+      }
+//      else {
+//
+//        q"""def $name[..$tparams](...$regularParamLists)(...$funImplParamLists): Strategy[$t] =
+//           new ${TypeName(className)}[..$targs](...$argLists)"""
+//      }
```
Comment (Bastacyclop) [inline macros/src/main/scala/elevate/macros/CombinatorMacro.scala]: do we need to keep this?

File: src/main/scala/elevate/core/strategies/debug.scala [hunk 2]
```diff
@@ -1,21 +1,16 @@
...
-    override def toString: String = "debug"
-  }
+  @rule def debug[P](msg: String): Strategy[P] =
+    peek[P](x => println(msg + "\n" + x))
 
-  case class echo[P](msg: String) extends Strategy[P] {
-    def apply(e: P): RewriteResult[P] = peek[P](_ => println(msg))(e)
-    override def toString: String = "println"
-  }
+  @rule("println") def echo[P](msg: String): Strategy[P] =
```
Comment (Bastacyclop) [inline src/main/scala/elevate/core/strategies/debug.scala]: "println" name for the "echo" strategy?

File: src/main/scala/elevate/rise/rules/algorithmic.scala [hunk 3]
```diff
@@ -27,294 +28,204 @@ object algorithmic {
...
-      Success((slide(u)(v) >> map(slide(n)(m)) >> join) :: expr.t)
-    case _ =>
-      Failure(slideOverlap(u))
+      Success((slide(u)(v) >> map(slide(n)(m)) >> join) :: e.t)
   }
 
   // slide widening
 
-  def dropInSlide: Strategy[Rise] =
-    `slide n 1 >> drop l -> slide (n+l) 1 >> map(drop l)`
```
Comment (Bastacyclop) [inline src/main/scala/elevate/rise/rules/algorithmic.scala:220]: I would like to keep these kind of documenting "example/specification" as a comment.

File: src/test/scala/elevate/fsmooth/rewriteExamples.scala [hunk 4]
```diff
@@ -64,7 +64,7 @@ class rewriteExamples extends elevate.test_util.Tests {
     println(step3)
 
     println("\n-- step4: partial evaluation + simplification")
-    val strategy3 = normalize.apply(letPartialEvaluation <+ additionZero)
+    val strategy3 = normalize(traversal.FSmoothTraversable)(letPartialEvaluation <+ additionZero)
```
Comment (Bastacyclop) [inline src/test/scala/elevate/fsmooth/rewriteExamples.scala:67]: the implicit does not work well anymore with currying I assume?

File: src/main/scala/elevate/core/strategies/predicate.scala [hunk 5]
```diff
@@ -13,40 +15,31 @@ object predicate {
...
 
-  case class isEqualTo[P](x: P) extends Strategy[P] {
-    def apply(p: P): RewriteResult[P] =
-      if (p == x) Success(p) else Failure(isEqualTo(x))
-    override def toString: String = s"isEqualTo($x)"
-  }
+  @strategy
+  def isEqualTo[P](x: P): Strategy[P] = p => if (p == x) Success(p) else Failure(isEqualTo(x))
 
   case class contains[P: Traversable](x: P) extends Strategy[P] {
```
Comment (bastian-koepcke) [inline src/main/scala/elevate/core/strategies/predicate.scala:38]: Shouldn't this be implemented using a macro as well?

File: src/test/scala/elevate/fsmooth/rewriteExamples.scala (hunk 4 above)
Comment (michel-steuwer) [reply src/test/scala/elevate/fsmooth/rewriteExamples.scala:67]: Yeah, they work differently (for now).

File: macros/src/main/scala/elevate/macros/CombinatorMacro.scala (hunk 1 above)
Comment (michel-steuwer) [reply macros/src/main/scala/elevate/macros/CombinatorMacro.scala]: I have removed these comments.

File: src/main/scala/elevate/core/strategies/debug.scala (hunk 2 above)
Comment (michel-steuwer) [reply src/main/scala/elevate/core/strategies/debug.scala]: I can't remember why I named this `println`, I have now removed the optional doc string.

File: src/main/scala/elevate/rise/rules/algorithmic.scala (hunk 3 above)
Comment (michel-steuwer) [reply src/main/scala/elevate/rise/rules/algorithmic.scala:220]: I added them back

File: src/main/scala/elevate/core/strategies/predicate.scala (hunk 5 above)
Comment (michel-steuwer) [reply src/main/scala/elevate/core/strategies/predicate.scala:38]: This shoiuld be added in the future.

**Review by bastian-koepcke — APPROVED** [review]

**Review by Bastacyclop — APPROVED** [review]

**Pull request merged**